File: 424-longest-repeating-character-replacement/424-longest-repeating-character-replacement.scala
import scala.collection.mutable._

object Solution {
    def characterReplacement(s: String, k: Int): Int = {
        var start = 0
        var end = 0
        val map: Map[Char, Int] = Map()
        var res = 0
        var maxCnt = 0

        for(end <- 0 to s.length - 1) {
            val c = s.charAt(end)
            val cnt = map.get(c).getOrElse(0) + 1
            map += (c -> cnt)

            maxCnt = maxCnt max cnt

            while(end - start + 1 - maxCnt > k) {
                val r = s.charAt(start)
                map += (r -> (map.get(r).get - 1))
                start += 1
            }

            res = res max (end - start + 1)
        }

        res
    }
}

File: 424-longest-repeating-character-replacement/424-longest-repeating-character-replacement.java
class Solution {
    public int characterReplacement(String s, int k) {
        int res = 0;
        int l = -1;
        
        int[] bucket = new int[26];
        
        for(int i = 0, con = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            
            con = Math.max(con, ++bucket[c - 'A']);
            
            if (i - l - con > k) {
                bucket[s.charAt(++l) - 'A']--;
            }
            
            res = Math.max(res, i - l);
        }
        
        return res;
    }
}